uaclient/util.py

```diff
@@ -754,9 +756,40 @@ def redact_sensitive_logs(
     return redacted_log
 
 
-def should_reboot() -> bool:
-    """Check if the system needs to be rebooted."""
-    return os.path.exists(REBOOT_FILE_CHECK_PATH)
+def should_reboot(installed_pkgs: Optional[Set[str]] = None) -> bool:
+    """Check if the system needs to be rebooted.
+
+       :param installed_pkgs: If provided, verify if the any packages in
+           the list are present on /var/run/reboot-required.pkgs. If that
+           param is provided, we will only return true if we have the
+           reboot-required marker file and any package in reboot-required.pkgs
+           file.
+    """
+
+    # If the reboot marker file doesn't exist, we don't even
+    # need to look at the installed_pkgs param
+    if not os.path.exists(REBOOT_FILE_CHECK_PATH):
+        return False
+
+    # If there is no installed_pkgs to check, we will rely only
+    # on the existence of the reboot marker file
+    if not installed_pkgs:
+        return True
+
+    try:
+        reboot_required_pkgs = set(
+            load_file(REBOOT_PKGS_FILE_PATH).split("\n")
+        )
+    except FileNotFoundError:
+        # If the file doesn't exist, we will default to the
+        # reboot  marker file
+        return True
+    else:
+        return (
+            True
+            if installed_pkgs.intersection(reboot_required_pkgs)
+            else False
+        )
```

Comment:
We can simplify this to just
```suggestion
        return installed_pkgs.intersection(reboot_required_pkgs)
```

Reply:
We can't make this change here because the `intersection` method will return another `Set`. Here, we just wan't to verify that the `Set` returned is non-empty.

However, we can update this call to:
```python
return len(installed_pkgs.intersection(reboot_required_pkg) != 0
```

Reply:
You both are totally right, any of those or the original would work, but not my suggestion haha :upside_down_face: